include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1898,6 +1898,80 @@ SparseMatrix<number>::get_sparsity_pattern () const
 }
 
 
+namespace internal
+{
+  namespace SparseMatrix
+  {
+    template <class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const realType&val)
+    {
+      out << val;
+    }
+
+    template<class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const std::complex<realType> &val)
+    {
+      // the following format is what is required by the matrix market format.
+      out << val.real << ' ' << val.imag;
+    }
+
+    template<class StreamType, class numberType>
+    void
+    print_header(StreamType &out,const numberType & dummy)
+    {
```

Comment:
Here, you use two template arguments that you let the compiler figure out by itself whenever you call the function as
```
  print_header (out, value);
```
That's ok, though I would avoid giving `dummy` a name since otherwise the compiler will complain about an unused argument. You can just write
```
  template <class StreamType, class numberType>
  void print_header (StreamType &out,const numberType & /*dummy*/)
```

But you don't actually use the value, and so one would think that there is a different way of doing this. Indeed, you could write the function as follows:
```
  template <typename realType> 
  struct PrintHeader 
  {
    template <typename StreamType> 
    static void print_value (StreamType &out)
    { ... }
  };

  template <typename realType> 
  struct PrintHeader<std::complex<realType> >   // implementation for complex scalars
  {
    template <typename StreamType> 
    static void print_value (StreamType &out)
    { ... }
  };
```
You would then just call the function as
```
  PrintHeader<number>::print_header (out);
```
and it would automatically dispatch to the right function.